lib/frame.c

```diff
@@ -71,7 +71,8 @@ int quicly_decode_ack_frame(const uint8_t **src, const uint8_t *end, quicly_ack_
         goto Error;
     if ((frame->ack_delay = quicly_decodev(src, end)) == UINT64_MAX)
         goto Error;
-    if ((frame->num_gaps = quicly_decodev(src, end)) == UINT64_MAX)
+    if ((frame->num_gaps = quicly_decodev(src, end)) == UINT64_MAX ||
+        frame->num_gaps > sizeof(frame->gaps) / sizeof(frame->gaps[0]))
```

Comment:
We cannot call it a frame encoding error when the number of gaps exceeds certain threshold, because that would be an attack vector for a middlebox to disrupt the connection. Consider the case where an on-path attacker drops every one of two packets.

IMO what we should do is decode an ACK frame of any number of gaps correctly, but silently cap the number of gaps (and ranges) that we keep in `quicly_ack_frame_t`. Then, the only negative effect would be us calling more packets as being lost than we should be.

Reply:
I see that the original fix was a bit eccentric. Thank you for pointing it out. Does 69cf480aacddb6d4124013cde68b9d550a557341 look more in line with your perspective?